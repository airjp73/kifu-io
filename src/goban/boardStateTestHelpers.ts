import { BoardState, GameState } from './gameStateReducer';

/**
 * A test helper to allow for creating board states for tests in a visual way.
 *
 * @param boardStateString A string representation of the board
 * @param size an array of 2 numbers representing the dimensions of the board
 */
export const createBoardStateFromString = (
  boardStateString: string,
  size: [number, number]
): BoardState => {
  const noWhitespace = boardStateString.replace(/\s/g, '');
  const A = 'a'.charCodeAt(0);

  const boardState: BoardState = {};
  for (let x = 0; x < size[0]; ++x) {
    for (let y = 0; y < size[1]; ++y) {
      const value = noWhitespace.charAt(y * size[0] + x);
      if (value === 'b' || value === 'w') {
        const point = String.fromCharCode(A + x) + String.fromCharCode(A + y);
        boardState[point] = value;
      }
    }
  }

  return boardState;
};

export const createStringFromBoardState = (
  gameState: GameState,
  size: [number, number]
): string => {
  const { boardState, captureCounts, properties } = gameState;
  const {
    playerBlack,
    playerWhite,
    teamBlack,
    teamWhite,
    rankBlack,
    rankWhite,
  } = properties;
  const playerString = (player: string, team: string, rank: string) =>
    `${player || team} ${rank}`;
  let boardString =
    '\n' +
    'Players:\n' +
    `Black: ${playerString(playerBlack, teamBlack, rankBlack)}\n` +
    `White: ${playerString(playerWhite, teamWhite, rankWhite)}\n\n` +
    'Captures:\n' +
    `Black: ${captureCounts.b}\n` +
    `White: ${captureCounts.w}\n\n` +
    'Board:\n';
  const A = 'a'.charCodeAt(0);

  for (let y = 0; y < size[1]; ++y) {
    for (let x = 0; x < size[0]; ++x) {
      const point = String.fromCharCode(x + A) + String.fromCharCode(y + A);
      const value = boardState[point];
      if (value === 'b' || value === 'w') {
        boardString += value + ' ';
      } else {
        boardString += '. ';
      }
    }
    boardString += '\n';
  }

  return boardString;
};
