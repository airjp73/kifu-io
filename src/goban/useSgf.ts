import { useMemo } from 'react';
import parseSgf from 'goban/parseSgf/parseSgf';
import normalizeGameTree, { GameTree } from 'goban/parseSgf/normalizeGameTree';

const useSgf = (sgf: string | null): [GameTree | null, Error | null] => {
  return useMemo(() => {
    if (!sgf) return [null, null];

    let gameTree: GameTree | null = null;
    let error: Error | null = null;

    try {
      gameTree = normalizeGameTree(parseSgf(sgf)[0]);
    } catch (err) {
      error = err;
    }
    return [gameTree, error];
  }, [sgf]);
};

export default useSgf;
