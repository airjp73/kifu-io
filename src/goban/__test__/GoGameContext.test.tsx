import React from 'react';
import each from 'jest-each';
import { render, fireEvent } from 'react-testing-library';
import sgf1 from 'goban/parseSgf/snapshots/snapshot1';
import sgf2 from 'goban/parseSgf/snapshots/snapshot2';
import sgf3 from 'goban/parseSgf/snapshots/snapshot3';
import useSgf from '../useSgf';
import { GoGameContextProvider, useGoGameContext } from '../GoGameContext';
import { createStringFromBoardState } from '../boardStateTestHelpers';

const BoardStateTestRenderer = () => {
  const { forward, back, gameState } = useGoGameContext();
  const boardString = createStringFromBoardState(
    gameState.boardState,
    gameState.properties.boardSize || [19, 19]
  );
  return (
    <div>
      <pre data-testid="boardString">{boardString}</pre>;
      <button data-testid="gotobeginning" onClick={() => back(-1)} />
      <button data-testid="back10" onClick={() => back(10)} />
      <button data-testid="back1" onClick={() => back(1)} />
      <button data-testid="forward1" onClick={() => forward(1)} />
      <button data-testid="forward10" onClick={() => forward(10)} />
      <button data-testid="gotoend" onClick={() => forward(-1)} />
    </div>
  );
};

interface TestGoGameProps {
  sgf: string;
}
const TestGoGame: React.FunctionComponent<TestGoGameProps> = ({ sgf }) => {
  const [gameTree] = useSgf(sgf);
  return (
    <GoGameContextProvider gameTree={gameTree}>
      <BoardStateTestRenderer />
    </GoGameContextProvider>
  );
};

describe('Game context snapshots', () => {
  each([
    ['sgf1', sgf1],
    ['sgf2', sgf2],
    ['sgf3', sgf3],
  ]).test('should render the final position correctly for %s', (desc, sgf) => {
    const game = render(<TestGoGame sgf={sgf} />);
    fireEvent.click(game.getByTestId('gotoend'));
    expect(game.getByTestId('boardString')).toMatchSnapshot();
  });

  // lists of moves generated randomly
  const sequences: string[][] = [
    [
      'forward10',
      'forward1',
      'forward1',
      'forward1',
      'forward10',
      'back1',
      'forward10',
      'forward1',
      'back10',
      'back10',
      'forward10',
      'back10',
      'forward10',
      'forward10',
      'back1',
      'forward1',
      'forward1',
      'forward1',
      'forward10',
      'back1',
    ],
    [
      'forward1',
      'forward10',
      'forward1',
      'back10',
      'forward10',
      'back10',
      'back1',
      'forward1',
      'forward1',
      'back10',
      'forward1',
      'forward10',
      'forward10',
      'gotoend',
      'forward10',
      'forward10',
      'back1',
      'forward10',
      'forward1',
      'forward10',
    ],
    [
      'back10',
      'back1',
      'gotobeginning',
      'gotobeginning',
      'back10',
      'gotobeginning',
      'forward10',
      'forward10',
      'forward10',
      'back10',
      'back1',
      'forward10',
      'forward1',
      'forward1',
      'forward10',
      'forward10',
      'forward10',
      'back10',
      'back10',
      'forward1',
    ],
    [
      'forward1',
      'gotoend',
      'back1',
      'gotoend',
      'forward1',
      'gotoend',
      'forward1',
      'back10',
      'back10',
      'gotoend',
      'forward1',
      'back10',
      'back10',
      'gotobeginning',
      'forward10',
      'back1',
      'forward10',
      'back1',
      'forward1',
      'back10',
    ],
    [
      'back1',
      'forward1',
      'forward1',
      'back10',
      'back1',
      'back1',
      'forward10',
      'back1',
      'forward10',
      'back1',
      'forward1',
      'back10',
      'back10',
      'gotobeginning',
      'forward10',
      'forward1',
      'forward10',
      'back1',
      'forward10',
      'back10',
      'gotobeginning',
      'forward10',
      'forward1',
      'forward10',
      'forward1',
      'back1',
      'forward10',
      'back1',
      'back10',
      'gotoend',
      'back10',
      'back1',
      'back1',
      'forward1',
      'back1',
      'forward10',
      'forward10',
      'back1',
      'gotobeginning',
      'forward10',
    ],
    [
      'back1',
      'back10',
      'forward10',
      'forward1',
      'back10',
      'forward1',
      'gotoend',
      'back1',
      'forward1',
      'forward1',
      'forward10',
      'forward10',
      'back10',
      'forward1',
      'back1',
      'forward10',
      'back1',
      'forward10',
      'back10',
      'back10',
      'back10',
      'forward1',
      'gotobeginning',
      'back10',
      'forward1',
      'forward10',
      'forward10',
      'forward1',
      'forward1',
      'forward10',
      'forward10',
      'gotobeginning',
      'back1',
      'back1',
      'back10',
      'gotobeginning',
      'forward1',
      'forward10',
      'forward10',
      'forward10',
    ],
  ];

  const cases = sequences
    .map((sequence, index) => [
      [index, 'sgf1', sgf1],
      [index, 'sgf2', sgf2],
      [index, 'sgf3', sgf3],
    ])
    // flatMap causes issues 🤷‍♀️
    .reduce((flattened, current) => [...flattened, ...current], []);

  each(cases).test(
    'should perform sequence %d correctly for %s',
    (sequenceIndex, desc, sgf) => {
      const sequence = sequences[sequenceIndex];
      const game = render(<TestGoGame sgf={sgf} />);
      sequence.forEach(testid => fireEvent.click(game.getByTestId(testid)));
      expect(game.getByTestId('boardString')).toMatchSnapshot();
    }
  );

  each([
    ['sgf1', sgf1],
    ['sgf2', sgf2],
    ['sgf3', sgf3],
  ]).test('should render move 50 correctly for %s', (desc, sgf) => {
    const game = render(<TestGoGame sgf={sgf} />);
    fireEvent.click(game.getByTestId('forward10'));
    fireEvent.click(game.getByTestId('forward10'));
    fireEvent.click(game.getByTestId('forward10'));
    fireEvent.click(game.getByTestId('forward10'));
    fireEvent.click(game.getByTestId('forward10'));
    expect(game.getByTestId('boardString')).toMatchSnapshot();
  });

  each([
    ['sgf1', sgf1],
    ['sgf2', sgf2],
    ['sgf3', sgf3],
  ]).test(
    'should render 50 moves from the end correctly for %s',
    (desc, sgf) => {
      const game = render(<TestGoGame sgf={sgf} />);
      fireEvent.click(game.getByTestId('gotoend'));
      fireEvent.click(game.getByTestId('back10'));
      fireEvent.click(game.getByTestId('back10'));
      fireEvent.click(game.getByTestId('back10'));
      fireEvent.click(game.getByTestId('back10'));
      fireEvent.click(game.getByTestId('back10'));
      expect(game.getByTestId('boardString')).toMatchSnapshot();
    }
  );
});
