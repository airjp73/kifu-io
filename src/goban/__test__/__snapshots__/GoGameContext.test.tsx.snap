// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game context snapshots should perform sequence 0 correctly for sgf1 1`] = `
"
Players:
Black: AaronP 2d
White: U undefined

Captures:
Black: 0
White: 0

Board:
. . . . . . . . . . . . . . . . . . . 
. . . . . . . . . . . . . . . . . . . 
. . . . . b w w w b . . . . . b w . . 
. . . w . w w b w b . . . . . b w . . 
. . . . . . . b b b . . . . . b w . . 
. . w . . . . . . . . . . . . . b w . 
. . . . . . . . . . . . . . . . b w . 
. . . . . . . . . . . . . . . . . . . 
. . . . . . . . . . . . . . . . b . . 
. . b . . . . . . . . . . . w b . . . 
. . . . . . . . . . . . . . b w w . . 
. . . . . . . . . . . . . . . . . . . 
. . . b . . . . . . . . . . . . . . . 
. . b . . . . . . . . . . . . . . . . 
. . w b . . . . . . . . . . . w w . . 
. . w . . . . . . . . . . . b . b . . 
. . . w . w . . . . . . . . . b . . . 
. . . . . . . . . . . . . . . . . . . 
. . . . . . . . . . . . . . . . . . . 
"
`;

exports[`Game context snapshots should perform sequence 0 correctly for sgf2 1`] = `
"
Players:
Black: Koons undefined
White: AaronP 2d

Captures:
Black: 1
White: 0

Board:
. . b . . . . . . . . . . . . . . . . 
. b b b . . . . . . . . . . . . . . . 
. w b w . b . . . . . . . w . w w w . 
. w b w . . . . . . . . . . b b b w . 
. w b w . . . . . . . . . . . . b b . 
. b w b w . . . . . . . . . . . w . . 
. w w . . . . . . . . . . . . . . . . 
. . . . . . . . . . . . . . . . . . . 
. . . . . . . . . . . . . . . . b . . 
. . . . . . . . . . . . . . . . . . . 
. . . . . . . . . . . . . . . . . . . 
. . b . . . . . . . . . . . . . . . . 
. . . . . . . . . . . . . . . . . . . 
. . . w . . . . . . . . . w . . . . . 
. . . . . . . . . . . . . . . . . . . 
. . . w . w . . b . . . . w b . b . . 
. . . . . b . . . . . . . . b . . . . 
. . . . . . . . . . . . . . . . . . . 
. . . . . . . . . . . . . . . . . . . 
"
`;

exports[`Game context snapshots should perform sequence 0 correctly for sgf3 1`] = `
"
Players:
Black: Peridothaze ?
White: AaronP 2k

Captures:
Black: 3
White: 2

Board:
. . . . b . . . . 
. . w b . b . . . 
b b b b . . b b . 
. w b b b b w b . 
w w w b w w w b . 
. w b b w . w b . 
. w b . . w b b . 
. w b . w . w b . 
. w . . . . w . . 
"
`;

exports[`Game context snapshots should perform sequence 1 correctly for sgf1 1`] = `
"
Players:
Black: AaronP 2d
White: U undefined

Captures:
Black: 19
White: 7

Board:
. . . . . . w b b b . b b b w . w . . 
. . . . . . . w b b b . . b w w . w . 
. . . . . b w w w b b b b w w b w . . 
. . . w w w w b w b . w w b b b w w . 
. w . w b w b b b b . w b . b b w b . 
w . w b b w b w . b . w b w . b b w . 
b w w b w w w w b . b b b . b b b w . 
b b w b b b b b . b w b b b w b w w w 
. b b . . . . b . b w b w w w b b b w 
. . b . . . . b b w w w w . w b b b b 
. . . . . . . . . b w . . b b w w w w 
. b . . . . . . . b w w w . b . w . . 
b . b b . . b b . w b w b b b b b w . 
w b b b b b . b w b b b b w w . w w . 
w w w b w w b b b w w . b w w w w . . 
. . w w . w b w b w w b b w b w . w . 
. . . w . w w w w w b b . b b b w . . 
. . . . . . . . . w w b . . b w w . . 
. . . . . . . . . w b b b b . w . . . 
"
`;

exports[`Game context snapshots should perform sequence 1 correctly for sgf2 1`] = `
"
Players:
Black: Koons undefined
White: AaronP 2d

Captures:
Black: 4
White: 6

Board:
. . b . . . . w . . w . . . . . . . . 
. b b b b w w . w w b w w . w . . . . 
. w b w w b w w b b b b b w b w w w . 
. w b w . b b b b . b b w w b b b w w 
. w b w . b . b . b w w w b . . b b b 
. b w w w b w . b w . w b . b . w w . 
. w w w b w w . b w w w b b . . . . . 
w . . w b . . . . b w b . w b b . b b 
b w . w b . . . . . b b . w w . b b w 
b b w w b . . b . . . . b b b b w w w 
. . b b b b b b b b b b w w w b w . . 
b b b . b w b w w w w w . b w w w w . 
b w w w b w w . w . w . . w b b b w w 
w w . w . w . w b b w . . w w b b b w 
. . . . w . w w b . w . . w b . b w w 
. . . w w w b b b . w w . w b . b b b 
. . . . w b . . . b w b w w b . . . . 
. . w . w b . b . b b b w b b . . . . 
. . . w b b . . . . b w w w b . . . . 
"
`;

exports[`Game context snapshots should perform sequence 1 correctly for sgf3 1`] = `
"
Players:
Black: Peridothaze ?
White: AaronP 2k

Captures:
Black: 3
White: 2

Board:
. . . . b . . . . 
. . w b . b . . . 
b b b b . . b b . 
b w b b b b w b . 
w w w b w w w b . 
. w b b w . w b . 
. w b b w w b b . 
. w b w w . w b b 
. w b b w . w w w 
"
`;

exports[`Game context snapshots should perform sequence 2 correctly for sgf1 1`] = `
"
Players:
Black: AaronP 2d
White: U undefined

Captures:
Black: 0
White: 0

Board:
. . . . . . . . . . . . . . . . . . . 
. . . . . . . . . . . . . . . . . . . 
. . . . . b w w w b . . . . . b w . . 
. . . w . w w b w b . . . . . b w . . 
. . . . . . . b b b . . . . . b w . . 
. . w . . . . . . . . . . . . . b w . 
. . . . . . . . . . . . . . . . b w . 
. . . . . . . . . . . . . . . . . . . 
. . . . . . . . . . . . . . . . b . . 
. . b . . . . . . . . . . . . b . . . 
. . . . . . . . . . . . . . . w w . . 
. . . . . . . . . . . . . . . . . . . 
. . . b . . . . . . . . . . . . . . . 
. . b . . . . . . . . . . . . . . . . 
. . w b . . . . . . . . . . . w w . . 
. . w . . . . . . . . . . . b . b . . 
. . . w . w . . . . . . . . . b . . . 
. . . . . . . . . . . . . . . . . . . 
. . . . . . . . . . . . . . . . . . . 
"
`;

exports[`Game context snapshots should perform sequence 2 correctly for sgf2 1`] = `
"
Players:
Black: Koons undefined
White: AaronP 2d

Captures:
Black: 1
White: 0

Board:
. . b . . . . . . . . . . . . . . . . 
. b b b . . . . . . . . . . . . . . . 
. w b w . b . . . . . . . w . w w w . 
. w b w . . . . . . . . . . b b b w . 
. w b w . . . . . . . . . . . . b b . 
. b w b w . . . . . . . . . . . w . . 
. w w . . . . . . . . . . . . . . . . 
. . . . . . . . . . . . . . . . . . . 
. . . . . . . . . . . . . . . . b . . 
. . . . . . . . . . . . . . . . . . . 
. . . . . . . . . . . . . . . . . . . 
. . . . . . . . . . . . . . . . . . . 
. . . . . . . . . . . . . . . . . . . 
. . . w . . . . . . . . . w . . . . . 
. . . . . . . . . . . . . . . . . . . 
. . . w . . . . b . . . . w b . b . . 
. . . . . b . . . . . . . . b . . . . 
. . . . . . . . . . . . . . . . . . . 
. . . . . . . . . . . . . . . . . . . 
"
`;

exports[`Game context snapshots should perform sequence 2 correctly for sgf3 1`] = `
"
Players:
Black: Peridothaze ?
White: AaronP 2k

Captures:
Black: 0
White: 1

Board:
. . . . b . . . . 
. . w b w b . . . 
. b b b w w b b . 
. w . b . . w b . 
w . w b w . w b . 
. w b b w b w b . 
. w b . . w b b . 
. w b . w . w . . 
. . . . . . . . . 
"
`;

exports[`Game context snapshots should perform sequence 3 correctly for sgf1 1`] = `
"
Players:
Black: AaronP 2d
White: U undefined

Captures:
Black: 0
White: 0

Board:
. . . . . . . . . . . . . . . . . . . 
. . . . . . . . . . . . . . . . . . . 
. . . . . b . . . . . . . . . . . . . 
. . . w . . . . . . . . . . . b . . . 
. . . . . . . . . . . . . . . . . . . 
. . . . . . . . . . . . . . . . . . . 
. . . . . . . . . . . . . . . . . . . 
. . . . . . . . . . . . . . . . . . . 
. . . . . . . . . . . . . . . . . . . 
. . . . . . . . . . . . . . . . . . . 
. . . . . . . . . . . . . . . . . . . 
. . . . . . . . . . . . . . . . . . . 
. . . . . . . . . . . . . . . . . . . 
. . b . . . . . . . . . . . . . . . . 
. . w b . . . . . . . . . . . . . . . 
. . w . . . . . . . . . . . . . . . . 
. . . w . . . . . . . . . . . b . . . 
. . . . . . . . . . . . . . . . . . . 
. . . . . . . . . . . . . . . . . . . 
"
`;

exports[`Game context snapshots should perform sequence 3 correctly for sgf2 1`] = `
"
Players:
Black: Koons undefined
White: AaronP 2d

Captures:
Black: 0
White: 0

Board:
. . . . . . . . . . . . . . . . . . . 
. . . . . . . . . . . . . . . . . . . 
. . . . . . . . . . . . . . . . . . . 
. . . w . . . . . . . . . . . b . . . 
. . . . . . . . . . . . . . . . . . . 
. . . . . . . . . . . . . . . . w . . 
. . . . . . . . . . . . . . . . . . . 
. . . . . . . . . . . . . . . . . . . 
. . . . . . . . . . . . . . . . b . . 
. . . . . . . . . . . . . . . . . . . 
. . . . . . . . . . . . . . . . . . . 
. . . . . . . . . . . . . . . . . . . 
. . . . . . . . . . . . . . . . . . . 
. . . w . . . . . . . . . . . . . . . 
. . . . . . . . . . . . . . . . . . . 
. . . w . . . . b . . . . . . . . . . 
. . . . . b . . . . . . . . b . . . . 
. . . . . . . . . . . . . . . . . . . 
. . . . . . . . . . . . . . . . . . . 
"
`;

exports[`Game context snapshots should perform sequence 3 correctly for sgf3 1`] = `
"
Players:
Black: Peridothaze ?
White: AaronP 2k

Captures:
Black: 0
White: 0

Board:
. . . . . . . . . 
. . w b w b . . . 
. . b . w w b b . 
. . . . . . w b . 
. . . . . . . . . 
. . . . . . . . . 
. . b . . . b . . 
. . . . . . . . . 
. . . . . . . . . 
"
`;

exports[`Game context snapshots should perform sequence 4 correctly for sgf1 1`] = `
"
Players:
Black: AaronP 2d
White: U undefined

Captures:
Black: 0
White: 0

Board:
. . . . . . . . . . . . . . . . . . . 
. . . . . . . . . . . . . . . . . . . 
. . . . . b . . . . . . . . . . . . . 
. . . w . . . . . . . . . . . b . . . 
. . . . . . . . . . . . . . . . . . . 
. . w . . . . . . . . . . . . . . . . 
. . . . . . . . . . . . . . . . . . . 
. . . . . . . . . . . . . . . . . . . 
. . . . . . . . . . . . . . . . . . . 
. . . . . . . . . . . . . . . . . . . 
. . . . . . . . . . . . . . . . . . . 
. . . . . . . . . . . . . . . . . . . 
. . . . . . . . . . . . . . . . . . . 
. . b . . . . . . . . . . . . . . . . 
. . w b . . . . . . . . . . . . . . . 
. . w . . . . . . . . . . . . . . . . 
. . . w . . . . . . . . . . . b . . . 
. . . . . . . . . . . . . . . . . . . 
. . . . . . . . . . . . . . . . . . . 
"
`;

exports[`Game context snapshots should perform sequence 4 correctly for sgf2 1`] = `
"
Players:
Black: Koons undefined
White: AaronP 2d

Captures:
Black: 0
White: 0

Board:
. . . . . . . . . . . . . . . . . . . 
. . . . . . . . . . . . . . . . . . . 
. . . . . . . . . . . . . . . . w . . 
. . . w . . . . . . . . . . . b . . . 
. . . . . . . . . . . . . . . . . . . 
. . . . . . . . . . . . . . . . w . . 
. . . . . . . . . . . . . . . . . . . 
. . . . . . . . . . . . . . . . . . . 
. . . . . . . . . . . . . . . . b . . 
. . . . . . . . . . . . . . . . . . . 
. . . . . . . . . . . . . . . . . . . 
. . . . . . . . . . . . . . . . . . . 
. . . . . . . . . . . . . . . . . . . 
. . . w . . . . . . . . . . . . . . . 
. . . . . . . . . . . . . . . . . . . 
. . . w . . . . b . . . . . . . . . . 
. . . . . b . . . . . . . . b . . . . 
. . . . . . . . . . . . . . . . . . . 
. . . . . . . . . . . . . . . . . . . 
"
`;

exports[`Game context snapshots should perform sequence 4 correctly for sgf3 1`] = `
"
Players:
Black: Peridothaze ?
White: AaronP 2k

Captures:
Black: 0
White: 0

Board:
. . . . . . . . . 
. . w b w b . . . 
. . b b w w b b . 
. . . . . . w b . 
. . . . . . . . . 
. . . . . . . . . 
. . b . . . b . . 
. . . . . . . . . 
. . . . . . . . . 
"
`;

exports[`Game context snapshots should perform sequence 5 correctly for sgf1 1`] = `
"
Players:
Black: AaronP 2d
White: U undefined

Captures:
Black: 0
White: 0

Board:
. . . . . . . . . . . . . . . . . . . 
. . . . . . . . . . . . . . . . . . . 
. . . . . b . w . . . . . . . b w . . 
. . . w . . . b . b . . . . . b w . . 
. . . . . . . . . . . . . . . b w . . 
. . w . . . . . . . . . . . . . b w . 
. . . . . . . . . . . . . . . . b w . 
. . . . . . . . . . . . . . . . . . . 
. . . . . . . . . . . . . . . . b . . 
. . b . . . . . . . . . . . . . . . . 
. . . . . . . . . . . . . . . . w . . 
. . . . . . . . . . . . . . . . . . . 
. . . b . . . . . . . . . . . . . . . 
. . b . . . . . . . . . . . . . . . . 
. . w b . . . . . . . . . . . w w . . 
. . w . . . . . . . . . . . b . b . . 
. . . w . w . . . . . . . . . b . . . 
. . . . . . . . . . . . . . . . . . . 
. . . . . . . . . . . . . . . . . . . 
"
`;

exports[`Game context snapshots should perform sequence 5 correctly for sgf2 1`] = `
"
Players:
Black: Koons undefined
White: AaronP 2d

Captures:
Black: 0
White: 0

Board:
. . . . . . . . . . . . . . . . . . . 
. b w . . . . . . . . . . . . . . . . 
. . b w . . . . . . . . . w . w w w . 
. . b w . . . . . . . . . . b b b w . 
. . b w . . . . . . . . . . . . b b . 
. b w . . . . . . . . . . . . . w . . 
. . . . . . . . . . . . . . . . . . . 
. . . . . . . . . . . . . . . . . . . 
. . . . . . . . . . . . . . . . b . . 
. . . . . . . . . . . . . . . . . . . 
. . . . . . . . . . . . . . . . . . . 
. . . . . . . . . . . . . . . . . . . 
. . . . . . . . . . . . . . . . . . . 
. . . w . . . . . . . . . w . . . . . 
. . . . . . . . . . . . . . . . . . . 
. . . w . . . . b . . . . w b . b . . 
. . . . . b . . . . . . . . b . . . . 
. . . . . . . . . . . . . . . . . . . 
. . . . . . . . . . . . . . . . . . . 
"
`;

exports[`Game context snapshots should perform sequence 5 correctly for sgf3 1`] = `
"
Players:
Black: Peridothaze ?
White: AaronP 2k

Captures:
Black: 0
White: 1

Board:
. . . . b . . . . 
. . w b w b . . . 
. b b b w w b b . 
. w . b . . w b . 
w . w b w . w b . 
. w b b . . w b . 
. w b . . w b b . 
. . . . w . w . . 
. . . . . . . . . 
"
`;

exports[`Game context snapshots should render 50 moves from the end correctly for sgf1 1`] = `
"
Players:
Black: AaronP 2d
White: U undefined

Captures:
Black: 9
White: 3

Board:
. . . . . . w . w . . b b b w . w . . 
. . . . . . . w b b b . . b w w . w . 
. . . . . b w w w b b b b w w b w . . 
. . . w . w w b w b . w w b b b w . . 
. w . w . . . b b b . w b . b b w . . 
. . w b . w . . . . . w b w . b b w . 
. w w b . w w . b w b b b . b b b w . 
. b w b . b b b . b w b b b w b w w . 
. b b . . . . b . b w b w w w b b . . 
. . b . . . . b b w w w w . w b b b b 
. . . . . . . . . b w . . b b w w w w 
. b . . . . . . . b w w w . b . . . . 
b . . b . . b b . w b w b b . b b w . 
w b b b b b . b w b b b . . . . . w . 
w w w b w w b b b w . . b w w w w . . 
. . w w . w b w b w . . b w b w . w . 
. . . w . w w w w . . b w b b b w . . 
. . . . . . . . . w . b w w b . w . . 
. . . . . . . . . . . . . b . . . . . 
"
`;

exports[`Game context snapshots should render 50 moves from the end correctly for sgf2 1`] = `
"
Players:
Black: Koons undefined
White: AaronP 2d

Captures:
Black: 3
White: 5

Board:
. . b . . . . w . . w . . . . . . . . 
. b b b b w w . w w b w w . . . . . . 
. w b w w b w w b b b b . w . w w w . 
. w b w . b b b b . . . w . b b b w . 
. w b w . b . b . b w . w b . . b b b 
. b w . w b w . b w . w b . b . w w . 
. w w w b w w . b w w . b b . . . . . 
. . . . b . . . . b . . . w b b . b b 
. w . w b . . . . . . b . w w . b b w 
. b . w b . . b . . . . b b b b . w w 
. . . . b . b . b . b . w w w b w . . 
b b b . b . . w . w w w . b w w w w . 
b w w . . . w . w . w . . . . b b w . 
w w . w . w . w b . . . . w . . b b w 
. . . . w . w w b . w . . . . . b w . 
. . . w w w b b b . w w . w b . b . . 
. . . . w b . . . b w b w w b . . . . 
. . w . w b . b . b b b w b b . . . . 
. . . w b b . . . . . . . . . . . . . 
"
`;

exports[`Game context snapshots should render 50 moves from the end correctly for sgf3 1`] = `
"
Players:
Black: Peridothaze ?
White: AaronP 2k

Captures:
Black: 0
White: 0

Board:
. . . . . . . . . 
. . . . . . . . . 
. . b . . w b b . 
. . . . . . w b . 
. . . . . . . . . 
. . . . . . . . . 
. . b . . . b . . 
. . . . . . . . . 
. . . . . . . . . 
"
`;

exports[`Game context snapshots should render move 50 correctly for sgf1 1`] = `
"
Players:
Black: AaronP 2d
White: U undefined

Captures:
Black: 0
White: 0

Board:
. . . . . . . . . . . . . . . . . . . 
. . . . . . . . . . . . . . . . . . . 
. . . . . b w w w b . . . . . b w . . 
. . . w . w w b w b . . . . . b w . . 
. . . . . . . b b b . . . . . b w . . 
. . w . . . . . . . . . . . . . b w . 
. . . . . . . . . . . . . . . . b w . 
. . . . . . . . . . . . . . . b . . . 
. . . . . . . . . . . . . . w . b . . 
. . b . . . . . . . . . w . w b . . . 
. . . . . . . . . . . . . . b w w . . 
. . . . . . . . . . . . . . b . . . . 
. . . b . . . . . . . . . . . . b w . 
. . b . . . . . . . . . . . . . . . . 
. . w b . . . . . . . . . . . w w . . 
. . w . . . . . . . . . . . b . b . . 
. . . w . w . . . . . . . . . b . . . 
. . . . . . . . . . . . . . . . . . . 
. . . . . . . . . . . . . . . . . . . 
"
`;

exports[`Game context snapshots should render move 50 correctly for sgf2 1`] = `
"
Players:
Black: Koons undefined
White: AaronP 2d

Captures:
Black: 1
White: 0

Board:
. . b . . . . . . . . . . . . . . . . 
. b b b . . . . . . . . . . . . . . . 
. w b w . b . . . . . . . w . w w w . 
. w b w . . . . . . . . . . b b b w . 
. w b w . . . . . . . . . . . . b b . 
. b w b w . . . . . . . . . . . w . . 
. w w . . . . . . . . . . . . . . . . 
. . . . . . . . . . . . . . . . . . . 
. . . . . . . . . . . . . . . . b . . 
. . . . . . . . . . . . . . . . . . . 
. . . . . . . . . . . . . . . . . . . 
. . b . . . . . . . . . . . . . . . . 
. . . . . . . . . . . . . . . . . . . 
. . . w . . . . . . . . . w . . . . . 
. . . . . b w . . . . . . . . . . . . 
. . . w w w b . b . . . . w b . b . . 
. . . . w b . . . . . . . . b . . . . 
. . . . . b . . . . . . . . . . . . . 
. . . . . . . . . . . . . . . . . . . 
"
`;

exports[`Game context snapshots should render move 50 correctly for sgf3 1`] = `
"
Players:
Black: Peridothaze ?
White: AaronP 2k

Captures:
Black: 3
White: 2

Board:
. . . . b . . . . 
. . w b . b . . . 
b b b b . . b b . 
. w b b b b w b . 
w w w b w w w b . 
. w b b w . w b . 
. w b b w w b b . 
. w b . w . w b b 
. w . b w . w w . 
"
`;

exports[`Game context snapshots should render the final position correctly for sgf1 1`] = `
"
Players:
Black: AaronP 2d
White: U undefined

Captures:
Black: 19
White: 7

Board:
. . . . . . w b b b . b b b w . w . . 
. . . . . . . w b b b . . b w w . w . 
. . . . . b w w w b b b b w w b w . . 
. . . w w w w b w b . w w b b b w w . 
. w . w b w b b b b . w b . b b w b . 
w . w b b w b w . b . w b w . b b w . 
b w w b w w w w b . b b b . b b b w . 
b b w b b b b b . b w b b b w b w w w 
. b b . . . . b . b w b w w w b b b w 
. . b . . . . b b w w w w . w b b b b 
. . . . . . . . . b w . . b b w w w w 
. b . . . . . . . b w w w . b . w . . 
b . b b . . b b . w b w b b b b b w . 
w b b b b b . b w b b b b w w . w w . 
w w w b w w b b b w w . b w w w w . . 
. . w w . w b w b w w b b w b w . w . 
. . . w . w w w w w b b . b b b w . . 
. . . . . . . . . w w b . . b w w . . 
. . . . . . . . . w b b b b . w . . . 
"
`;

exports[`Game context snapshots should render the final position correctly for sgf2 1`] = `
"
Players:
Black: Koons undefined
White: AaronP 2d

Captures:
Black: 4
White: 6

Board:
. . b . . . . w . . w . . . . . . . . 
. b b b b w w . w w b w w . w . . . . 
. w b w w b w w b b b b b w b w w w . 
. w b w . b b b b . b b w w b b b w w 
. w b w . b . b . b w w w b . . b b b 
. b w w w b w . b w . w b . b . w w . 
. w w w b w w . b w w w b b . . . . . 
w . . w b . . . . b w b . w b b . b b 
b w . w b . . . . . b b . w w . b b w 
b b w w b . . b . . . . b b b b w w w 
. . b b b b b b b b b b w w w b w . . 
b b b . b w b w w w w w . b w w w w . 
b w w w b w w . w . w . . w b b b w w 
w w . w . w . w b b w . . w w b b b w 
. . . . w . w w b . w . . w b . b w w 
. . . w w w b b b . w w . w b . b b b 
. . . . w b . . . b w b w w b . . . . 
. . w . w b . b . b b b w b b . . . . 
. . . w b b . . . . b w w w b . . . . 
"
`;

exports[`Game context snapshots should render the final position correctly for sgf3 1`] = `
"
Players:
Black: Peridothaze ?
White: AaronP 2k

Captures:
Black: 3
White: 2

Board:
. . . . b . . . . 
. . w b . b . . . 
b b b b . . b b . 
b w b b b b w b . 
w w w b w w w b . 
. w b b w . w b . 
. w b b w w b b . 
. w b w w . w b b 
. w b b w . w w w 
"
`;
